File: content/bdd-with-cucumber/07/code/ruby/09-define-step/features/step_definitions/steps.rb
require 'shouty'

DEFAULT_RANGE = 100

Before do
  @people = {}
  @network = Shouty::Network.new(DEFAULT_RANGE)
  @messages_shouted_by = Hash.new([])
end

Given "the range is {int}" do |range|
  @network = Shouty::Network.new(range)
end

Given "a person named {word}" do |name|
  @people[name] = Shouty::Person.new(@network, 0)
end

Given "people are located at" do |table|
  table.transpose.symbolic_hashes.each do |name: , location: |
    @people[name] = Shouty::Person.new(@network, location.to_i)
  end
end

Given('Sean has bought {int} credits') do |credits|
  @people["Sean"].credits = credits
end

When "Sean shouts" do
  @people["Sean"].shout("Hello, world")
  @messages_shouted_by["Sean"] << "Hello, world"
end

When "Sean shouts {string}" do |message|
  @people["Sean"].shout(message)
  @messages_shouted_by["Sean"] << message
end

When 'Sean shouts the following message' do |message|
  @people["Sean"].shout(message)
  @messages_shouted_by["Sean"] << message
end

When "Sean shouts a message" do
  message = "A message from Sean"
  @people["Sean"].shout(message)
  @messages_shouted_by["Sean"] << message
end

When "Sean shouts a long message" do
  message = ["A message from Sean", "that spans multiple lines"].join("\n")
  @people["Sean"].shout(message)
  @messages_shouted_by["Sean"] << message
end

When 'Sean shouts a message containing the word {string}' do |word|
  message = "A message containing the word #{word}"
  @people["Sean"].shout(message)
  @messages_shouted_by["Sean"] << message
end

Then "Lucy should hear Sean's message" do
  expect(@people['Lucy'].messages_heard).to eq [@messages_shouted_by["Sean"][0]]
end

Then "Lucy should hear a shout" do
  expect(@people['Lucy'].messages_heard.count).to eq 1
end

Then "Larry should not hear Sean's message" do
  expect(@people['Larry'].messages_heard).not_to include(@messages_shouted_by["Sean"][0])
end

Then "{word} should not hear a shout" do |name|
  expect(@people[name].messages_heard.count).to eq 0
end

Then "Lucy hears the following messages:" do |expected_messages|
  actual_messages = @people['Lucy'].messages_heard.map { |message| [ message ] }

  expected_messages.diff!(actual_messages)
end

Then("Lucy hears all Sean's messages") do
  expect(@people["Lucy"].messages_heard).to match(@messages_shouted_by["Sean"])
end

Then("Sean should have {int} credits") do |credits|
  expect(@people["Sean"].credits).to eql(credits)
end

File: content/bdd-with-cucumber/07/code/js/05-use-intention-revealing-step-everywhere/features/step_definitions/steps.js
const { Given, When, Then, Before } = require("@cucumber/cucumber")
const { assertThat, contains, is, not, containsInAnyOrder } = require("hamjest")
const assert = require("assert")

const { Person, Network } = require("../../src/shouty")

const default_range = 100

Before(function () {
  this.people = {}
  this.messagesShoutedBy = {}
  this.network = new Network(default_range)
})

Given("the range is {int}", function (range) {
  this.network = new Network(range)
})

Given("a person named {word}", function (name) {
  this.people[name] = new Person(this.network, 0)
})

Given("people are located at", function (dataTable) {
  dataTable
    .transpose()
    .hashes()
    .map((person) => {
      this.people[person.name] = new Person(this.network, person.location)
    })
})

Given("Sean has bought {int} credits", function (credits) {
  this.people["Sean"].credits = credits
})

When("Sean shouts", function () {
  this.people["Sean"].shout("Hello, world")
})

When("Sean shouts a message containing the word {string}", function (word) {
  const message = `A message containing the word ${word}`
  this.people["Sean"].shout(message)
  if (!this.messagesShoutedBy["Sean"]) this.messagesShoutedBy["Sean"] = []
  this.messagesShoutedBy["Sean"].push(message)
})

When("Sean shouts {string}", function (message) {
  this.people["Sean"].shout(message)
  if (!this.messagesShoutedBy["Sean"]) this.messagesShoutedBy["Sean"] = []
  this.messagesShoutedBy["Sean"].push(message)
})

When("Sean shouts the following message", function (message) {
  this.people["Sean"].shout(message)
  if (!this.messagesShoutedBy["Sean"]) this.messagesShoutedBy["Sean"] = []
  this.messagesShoutedBy["Sean"].push(message)
})

Then("Lucy should hear Sean's message", function () {
  assertThat(this.messagesShoutedBy["Sean"].length, is(1))
  const message = this.messagesShoutedBy["Sean"][0]
  assertThat(this.people["Lucy"].messagesHeard(), contains(message))
})

Then("Lucy should hear a shout", function () {
  assertThat(this.people["Lucy"].messagesHeard().length, is(1))
})

Then("Larry should not hear Sean's message", function () {
  assertThat(this.messagesShoutedBy["Sean"].length, is(1))
  const message = this.messagesShoutedBy["Sean"][0]
  assertThat(this.people["Larry"].messagesHeard(), not(contains(message)))
})

Then("{word} should not hear a shout", function (name) {
  assertThat(this.people[name].messagesHeard().length, is(0))
})

Then("Lucy hears the following messages:", function (expectedMessages) {
  let actualMessages = this.people["Lucy"]
    .messagesHeard()
    .map((message) => [message])

  assert.deepEqual(actualMessages, expectedMessages.raw())
})

Then("Lucy hears all Sean's messages", function () {
  assert.deepEqual(
    this.people["Lucy"].messagesHeard(),
    this.messagesShoutedBy["Sean"]
  )
})

Then("Sean should have {int} credits", function (expectedCredits) {
  assertThat(this.people["Sean"].credits, is(expectedCredits))
})
